File: examples/cicadascript/vect.cs
union vect_t {
  null_vect_t
  cons_vect_t
} {
  t: type
  length: nat_t
}

class null_vect_t {
  t: type
  length: nat_t = zero_t ()
}

class cons_vect_t {
  {{ n: nat_t }}
  t: type
  length: nat_t = succ_t (this.n)
  car: this.t
  cdr: vect_t (this.t, this.n)
}

vect_append: (
  {{ t: type, m: nat_t, n: nat_t }}
  ante: vect_t (t, m),
  succ: vect_t (t, n),
) -> vect_t (t, nat_add (m, n)) = case (ante) {
  null_vect_t => succ
  cons_vect_t => cons_vect_t (
    car = ante.car
    cdr = vect_append (ante.cdr, succ)
  )
}

vect_map: (
  {{ a: type, b: type, n: nat_t }}
  fun: (a) -> b,
  vect: vect_t (a, n),
) -> vect_t (a, n) = case (vect) {
  null_vect_t => vect
  cons_vect_t => cons_vect_t (
    car = fun (vect.car)
    cdr = vect_map (fun, vect.cdr)
  )
}

File: src/cicada/std/vect.ts
import * as ut from "../../util"
import * as cc from "../core"
import { type_t } from "../type-of-type"
import { ref_t } from "../ref"
import { nat } from "./nat"

export
function vect (): cc.module_t {
  let m = new cc.module_t ("vect")
//   m.use (nat ())
//   m.define ("vect_t", new union_t ("vect_t", {
//     vect_null_t: new ref_t (m, "vect_null_t"),
//     vect_cons_t: new ref_t (m, "vect_cons_t"),
//   }))
//   m.define ("vect_null_t", new record_t ("vect_null_t", {
//     t: new type_t (),
//     length: new ref_t (m, "zero_t"),
//   }))
//   m.define ("vect_cons_t", new record_t ("vect_cons_t", {
//     [implicit]: {
//       n: new ref_t (m, "nat_t"),
//     },
//     t: new type_t (),
//     length: new record_t ("succ_t", {
//       prev: new dot_t ("this", "n"),
//       prev: new this_t ("n"),
//     }),
//     car: new dot_t ("this", "t"),
//     cdr: new record_t ("vect_t", {
//       t: new dot_t ("this", "t"),
//       length: new dot_t ("this", "n"),
//     }),
//   }))
  return m
}
